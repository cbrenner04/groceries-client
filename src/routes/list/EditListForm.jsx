import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import * as $ from 'jquery';

import * as config from '../../config/default';
import Alert from '../../components/Alert';
import { SelectField, TextField, CheckboxField } from '../../components/FormFields';
import { setUserInfo } from '../../utils/auth';

function EditListForm(props) {
  const [id, setId] = useState(0);
  const [errors, setErrors] = useState('');
  const [name, setName] = useState('');
  const [completed, setCompleted] = useState(false);
  const [type, setType] = useState('GroceryList');

  useEffect(() => {
    if (!props.match) props.history.push('/lists');
    $.ajax({
      type: 'GET',
      url: `${config.apiBase}/lists/${props.match.params.id}/edit`,
      dataType: 'JSON',
      headers: JSON.parse(sessionStorage.getItem('user')),
    }).done(({ list, current_user_id: currentUserId }, _status, request) => {
      setUserInfo(request);
      if (list.owner_id !== currentUserId) props.history.push('/lists');
      setId(list.id);
      setName(list.name);
      setCompleted(list.completed);
      setType(list.type);
    });
  }, [props.history, props.match]);

  const handleSubmit = (event) => {
    event.preventDefault();
    const list = {
      name,
      completed,
      type,
    };
    $.ajax({
      url: `${config.apiBase}/lists/${id}`,
      data: { list },
      method: 'PUT',
      headers: JSON.parse(sessionStorage.getItem('user')),
    }).done((_data, _status, request) => {
      setUserInfo(request);
      props.history.push('/lists');
    }).fail((response) => {
      const responseJSON = JSON.parse(response.responseText);
      const returnedErrors = Object.keys(responseJSON).map(key => `${key} ${responseJSON[key]}`);
      setErrors(returnedErrors.join(' and '));
    });
  };

  return (
    <div>
      <h1>Edit { name }</h1>
      <Link to="/lists" className="pull-right">
        Back to lists
      </Link>
      <br />
      <Alert errors={errors} handleDismiss={() => setErrors('')} />
      <form className="form" onSubmit={handleSubmit} autoComplete="off">
        <TextField
          name="name"
          label="Name"
          value={name}
          handleChange={({ target: { value } }) => setName(value)}
        />
        <SelectField
          name="type"
          label="Type"
          value={type}
          handleChange={({ target: { value } }) => setType(value)}
          options={[
            { value: 'BookList', label: 'books' },
            { value: 'GroceryList', label: 'groceries' },
            { value: 'MusicList', label: 'music' },
            { value: 'ToDoList', label: 'to-do' },
          ]}
          blankOption={false}
        />
        <CheckboxField
          name="completed"
          label="Completed"
          value={completed}
          handleChange={() => setCompleted(!completed)}
          blankOption={false}
          classes="mb-3"
        />
        <button type="submit" className="btn btn-success btn-block">
          Update List
        </button>
      </form>
    </div>
  );
}

EditListForm.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
      list_id: PropTypes.string,
    }).isRequired,
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func,
  }).isRequired,
};

export default EditListForm;
