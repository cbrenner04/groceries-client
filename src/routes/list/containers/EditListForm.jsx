import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Button, Form } from 'react-bootstrap';

import Alert from '../../../components/Alert';
import { SelectField, TextField, CheckboxField } from '../../../components/FormFields';
import axios from '../../../utils/api';

function EditListForm(props) {
  const [errors, setErrors] = useState('');
  const [name, setName] = useState(props.name);
  const [completed, setCompleted] = useState(props.completed);
  const [type, setType] = useState(props.type);

  const handleSubmit = async event => {
    event.preventDefault();
    const list = {
      name,
      completed,
      type,
    };
    try {
      await axios.put(`/lists/${props.listId}`, { list });
      props.history.push('/lists');
    } catch ({ response, request, message }) {
      if (response) {
        if (response.status === 401) {
          // TODO: how do we pass error messages along?
          props.history.push('/users/sign_in');
        } else if (response.status === 403) {
          // TODO: how do we pass error messages along?
          props.history.push('/lists');
        } else {
          const keys = Object.keys(response.data);
          const responseErrors = keys.map(key => `${key} ${response.data[key]}`);
          setErrors(responseErrors.join(' and '));
        }
      } else if (request) {
        // TODO: what do here?
      } else {
        setErrors(message);
      }
    }
  };

  return (
    <>
      <h1>Edit {name}</h1>
      <Button href="/lists" className="float-right" variant="link">
        Back to lists
      </Button>
      <br />
      <Alert errors={errors} handleDismiss={() => setErrors('')} />
      <Form onSubmit={handleSubmit} autoComplete="off">
        <TextField name="name" label="Name" value={name} handleChange={({ target: { value } }) => setName(value)} />
        <SelectField
          name="type"
          label="Type"
          value={type}
          handleChange={({ target: { value } }) => setType(value)}
          options={[
            { value: 'BookList', label: 'books' },
            { value: 'GroceryList', label: 'groceries' },
            { value: 'MusicList', label: 'music' },
            { value: 'ToDoList', label: 'to-do' },
          ]}
          blankOption={false}
        />
        <CheckboxField
          name="completed"
          label="Completed"
          value={completed}
          handleChange={() => setCompleted(!completed)}
          blankOption={false}
          classes="mb-3"
        />
        <Button type="submit" variant="success" block>
          Update List
        </Button>
      </Form>
    </>
  );
}

EditListForm.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func,
  }),
  listId: PropTypes.number,
  name: PropTypes.string,
  type: PropTypes.string,
  completed: PropTypes.bool,
};

export default EditListForm;
