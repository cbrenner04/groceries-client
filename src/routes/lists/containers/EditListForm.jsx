import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Form } from 'react-bootstrap';
import { toast } from 'react-toastify';

import ListFormFields from '../components/ListFormFields';
import axios from '../../../utils/api';
import FormSubmission from '../../../components/FormSubmission';

function EditListForm(props) {
  const [name, setName] = useState(props.name);
  const [completed, setCompleted] = useState(props.completed);
  const [type, setType] = useState(props.type);

  const handleSubmit = async (event) => {
    event.preventDefault();
    const list = {
      name,
      completed,
      type,
    };
    try {
      await axios.put(`/lists/${props.listId}`, { list });
      toast('List successfully updated', { type: 'info' });
      props.history.push('/lists');
    } catch ({ response, request, message }) {
      if (response) {
        if (response.status === 401) {
          toast('You must sign in', { type: 'error' });
          props.history.push('/users/sign_in');
        } else if ([403, 404].includes(response.status)) {
          toast('List not found', { type: 'error' });
          props.history.push('/lists');
        } else {
          const keys = Object.keys(response.data);
          const responseErrors = keys.map((key) => `${key} ${response.data[key]}`);
          toast(responseErrors.join(' and '), { type: 'error' });
        }
      } else if (request) {
        toast('Something went wrong', { type: 'error' });
      } else {
        toast(message, { type: 'error' });
      }
    }
  };

  return (
    <>
      <h1>Edit {name}</h1>
      <br />
      <Form onSubmit={handleSubmit} autoComplete="off">
        <ListFormFields
          name={name}
          type={type}
          completed={completed}
          handleNameChange={({ target: { value } }) => setName(value)}
          handleTypeChange={({ target: { value } }) => setType(value)}
          handleCompletedChange={() => setCompleted(!completed)}
          editForm
        />
        <FormSubmission
          submitText="Update List"
          cancelAction={() => props.history.push('/lists')}
          cancelText="Cancel"
        />
      </Form>
    </>
  );
}

EditListForm.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  listId: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  completed: PropTypes.bool.isRequired,
};

export default EditListForm;
