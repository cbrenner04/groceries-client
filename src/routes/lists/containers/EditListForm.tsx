import React, { type ChangeEvent, type FormEvent, useState } from 'react';
import { Form } from 'react-bootstrap';
import { toast } from 'react-toastify';
import { useNavigate } from 'react-router-dom';
import { type AxiosError } from 'axios';

import axios from 'utils/api';
import FormSubmission from 'components/FormSubmission';

import ListFormFields from '../components/ListFormFields';

export interface IEditListFormProps {
  listId: string;
  name: string;
  type: string;
  completed: boolean;
}

const EditListForm: React.FC<IEditListFormProps> = (props): React.JSX.Element => {
  const [name, setName] = useState(props.name);
  const [completed, setCompleted] = useState(props.completed);
  const [type, setType] = useState(props.type);
  const navigate = useNavigate();

  const handleSubmit = async (event: FormEvent<HTMLFormElement>): Promise<void> => {
    event.preventDefault();
    const list = {
      name,
      completed,
      type,
    };
    try {
      await axios.put(`/lists/${props.listId}`, { list });
      toast('List successfully updated', { type: 'info' });
      navigate('/lists');
    } catch (err: unknown) {
      const error = err as AxiosError;
      if (error.response) {
        if (error.response.status === 401) {
          toast('You must sign in', { type: 'error' });
          navigate('/users/sign_in');
        } else if ([403, 404].includes(error.response.status)) {
          toast('List not found', { type: 'error' });
          navigate('/lists');
        } else {
          const keys = Object.keys(error.response.data!);
          const responseErrors = keys.map((key) => `${key} ${(error.response?.data as Record<string, string>)[key]}`);
          toast(responseErrors.join(' and '), { type: 'error' });
        }
      } else if (error.request) {
        toast('Something went wrong', { type: 'error' });
      } else {
        toast(error.message, { type: 'error' });
      }
    }
  };

  return (
    <React.Fragment>
      <h1>Edit {name}</h1>
      <br />
      <Form onSubmit={handleSubmit} autoComplete="off">
        <ListFormFields
          name={name}
          type={type}
          completed={completed}
          handleNameChange={({ target: { value } }: ChangeEvent<HTMLInputElement>): void => setName(value)}
          handleTypeChange={({ target: { value } }: ChangeEvent<HTMLInputElement>): void => setType(value)}
          handleCompletedChange={(): void => setCompleted(!completed)}
          editForm
        />
        <FormSubmission
          submitText="Update List"
          cancelAction={(): void => navigate('/lists')}
          cancelText="Cancel"
          displayCancelButton={true}
        />
      </Form>
    </React.Fragment>
  );
};

export default EditListForm;
