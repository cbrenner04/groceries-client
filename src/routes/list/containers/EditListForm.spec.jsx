import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { toast } from 'react-toastify';

import EditListForm from './EditListForm';
import axios from '../../../utils/api';

jest.mock('react-toastify', () => ({
  toast: jest.fn(),
}));

describe('EditListForm', () => {
  const history = createMemoryHistory();
  const props = {
    history,
    listId: 1,
    name: 'foo',
    type: 'GroceryList',
    completed: false,
  };
  const renderEditListForm = (props) => {
    return render(
      <Router history={history}>
        <EditListForm {...props} />)
      </Router>,
    );
  };

  it('renders', () => {
    const { container } = renderEditListForm(props);

    expect(container).toMatchSnapshot();
  });

  it('updates name when changed', async () => {
    const { getByLabelText } = renderEditListForm(props);

    fireEvent.change(getByLabelText('Name'), { target: { value: 'a' } });

    expect(getByLabelText('Name')).toHaveValue('a');
  });

  it('updates type when changed', async () => {
    const { getByLabelText } = renderEditListForm(props);

    expect(getByLabelText('Type')).toHaveValue('GroceryList');

    fireEvent.change(getByLabelText('Type'), { target: { value: 'BookList' } });

    expect(getByLabelText('Type')).toHaveValue('BookList');
  });

  it('updates completed when changed', async () => {
    const { getByLabelText } = renderEditListForm(props);

    fireEvent.click(getByLabelText('Completed'));

    expect(getByLabelText('Completed')).toBeChecked();
  });

  it('makes post, displays toast, and redirects to lists page on successful submission', async () => {
    const data = {
      foo: 'bar',
    };
    axios.put = jest.fn().mockResolvedValue({ data });
    const { getByRole } = renderEditListForm(props);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('List successfully updated', { type: 'info' });
    expect(history.location.pathname).toBe('/lists');
  });

  it('redirects to user login when 401', async () => {
    axios.put = jest.fn().mockRejectedValue({ response: { status: 401 } });
    const { getByRole } = renderEditListForm(props);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('You must sign in', { type: 'error' });
    expect(history.location.pathname).toBe('/users/sign_in');
  });

  it('redirects to lists page when 403', async () => {
    axios.put = jest.fn().mockRejectedValue({ response: { status: 403 } });
    const { getByRole } = renderEditListForm(props);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('List not found', { type: 'error' });
    expect(history.location.pathname).toBe('/lists');
  });

  it('redirects to lists page when 404', async () => {
    axios.put = jest.fn().mockRejectedValue({ response: { status: 404 } });
    const { getByRole } = renderEditListForm(props);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('List not found', { type: 'error' });
    expect(history.location.pathname).toBe('/lists');
  });

  it('displays appropriate error message', async () => {
    axios.put = jest.fn().mockRejectedValue({
      response: {
        status: 500,
        data: {
          foo: 'bar',
          baz: 'foobar',
        },
      },
    });

    const { getByRole } = renderEditListForm(props);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('foo bar and baz foobar', { type: 'error' });
  });

  it('displays toast when error in sending request', async () => {
    axios.put = jest.fn().mockRejectedValue({
      request: 'request failed',
    });

    const { getByRole } = renderEditListForm(props);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('Something went wrong', { type: 'error' });
  });

  it('displays toast when unknown error', async () => {
    axios.put = jest.fn().mockRejectedValue({
      message: 'request failed',
    });

    const { getByRole } = renderEditListForm(props);

    fireEvent.click(getByRole('button'));
    await waitFor(() => expect(axios.put).toHaveBeenCalledTimes(1));

    expect(toast).toHaveBeenCalledWith('request failed', { type: 'error' });
  });
});
