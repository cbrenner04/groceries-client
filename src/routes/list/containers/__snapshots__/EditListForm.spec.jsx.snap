// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditListForm renders 1`] = `
<div>
  <h1>
    Edit 
    foo
  </h1>
  <a
    class="float-right"
    href="/lists"
  >
    Back to lists
  </a>
  <br />
  <form
    autocomplete="off"
    class=""
  >
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="name"
      >
        Name
      </label>
      <input
        class="form-control"
        id="name"
        name="name"
        placeholder=""
        type="text"
        value="foo"
      />
    </div>
    <div
      class="form-group"
    >
      <label
        class="form-label"
        for="type"
      >
        Type
      </label>
      <select
        class="form-control"
        id="type"
        name="type"
      >
        <option
          value="BookList"
        >
          books
        </option>
        <option
          value="GroceryList"
        >
          groceries
        </option>
        <option
          value="MusicList"
        >
          music
        </option>
        <option
          value="ToDoList"
        >
          to-do
        </option>
      </select>
    </div>
    <div
      class="mb-3 form-group"
    >
      <div
        class="form-check"
      >
        <input
          class="form-check-input"
          id="completed"
          name="completed"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="completed"
          title=""
        >
          Completed
        </label>
      </div>
    </div>
    <button
      class="btn btn-success btn-block"
      type="submit"
    >
      Update List
    </button>
  </form>
  )
</div>
`;
